lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs

```diff
@@ -287,24 +400,382 @@ withNetworkLayer tr network np project k =
+
+            getAddressTxs
+                :: BlockHeader
+                -> BlockHeader
+                -> Either Address RewardAccount
+                -> IO ChainEvents
+            getAddressTxs bhFrom bhTo = runBF . fmap fromBlockEvents . \case
+                Left address -> do
+                    txs <- BF.allPages \paged -> BF.getAddressTransactions'
+                        (BF.Address (encodeAddress @nd address))
+                        paged
+                        BF.Ascending
+                        (Just $ headerToIndex bhFrom)
+                        (Just $ headerToIndex bhTo)
+                    forConcurrently txs \BF.AddressTransaction{..} -> do
+                        (bftx, tx) <-
+                            fetchTransaction network _addressTransactionTxHash
+                        txIndex <-
+                            _addressTransactionTxIndex
+                                <?#> "_addressTransactionTxIndex"
+                        txBlockEvents
+                            bftx
+                            (unsafeMkSublist [(txIndex, tx)])
+                            (wholeList [])
+                Right account -> do
+                    let address = BF.Address $ encodeStakeAddress @nd account
+                    rs <- BF.getAccountRegistrations address <&> fmap \reg ->
+                        ( BF._accountRegistrationTxHash reg
+                        , pure case BF._accountRegistrationAction reg of
+                            BF.Registered -> CertRegisterKey account
+                            BF.Deregistered -> CertDelegateNone account
+                        )
+                    ds <- BF.getAccountDelegations address >>= traverse \d -> do
+                        pool <- fromBlockfrostM $ BF._accountDelegationPoolId d
+                        pure ( BF._accountDelegationTxHash d
+                             , [ CertDelegateFull account pool ]
+                             )
+                    blockEventsRegDeleg <-
+                        forConcurrently (rs <> ds) \(txHash, dcerts) -> do
+                        tx@BF.Transaction{_transactionIndex} <- BF.getTx txHash
+                        txIndex <- _transactionIndex <?#> "_transactionIndex"
+                        txBlockEvents
+                            tx
+                            (wholeList [])
+                            (unsafeMkSublist $ (txIndex,) <$> dcerts)
+                    ws <- BF.getAccountWithdrawals address
+                    blockEventsWithdraw <-
+                        forConcurrently ws \BF.AccountWithdrawal{..} -> do
+                        (bftx@BF.Transaction{_transactionIndex}, tx) <-
+                            fetchTransaction network _accountWithdrawalTxHash
+                        txIndex <- _transactionIndex <?#> "_transactionIndex"
+                        txBlockEvents
+                            bftx
+                            (unsafeMkSublist [(txIndex, tx)])
```

Comment:
Sorting may be required here as well. Sorry for not mentioning that `unsafe` fact. 😱

Reply:
Its a singleton here too, we're good.